vm_backtrace.c

```diff
@@ -783,6 +791,36 @@ rb_backtrace_to_location_ary(VALUE self)
     return bt->locary;
 }
 
+VALUE
+rb_location_ary_to_backtrace(VALUE ary)
+{
+    if (!RB_TYPE_P(ary, T_ARRAY) || !rb_frame_info_p(RARRAY_AREF(ary, 0))) {
+        return Qfalse;
+    }
+
+    rb_backtrace_t *new_backtrace;
+    long num_frames = RARRAY_LEN(ary);
+    VALUE btobj = backtrace_alloc_capa(num_frames, &new_backtrace);
+
+    for (long index = 0; index < RARRAY_LEN(ary); index++) {
+        VALUE locobj = RARRAY_AREF(ary, index);
+        struct valued_frame_info *src_vloc;
+        TypedData_Get_Struct(locobj, struct valued_frame_info, &location_data_type, src_vloc);
+
+        rb_backtrace_location_t *dst_location = &new_backtrace->backtrace[index];
+        RB_OBJ_WRITE(btobj, &dst_location->cme, src_vloc->loc->cme);
+        RB_OBJ_WRITE(btobj, &dst_location->iseq, src_vloc->loc->iseq);
+        dst_location->pc = src_vloc->loc->pc;
+
+        new_backtrace->backtrace_size++;
+
+        RB_GC_GUARD(locobj);
+    }
+
+    RB_GC_GUARD(btobj);
```

Comment:
I think it's unnecessary to guard this if you're returning it since `btobj` obviously must either be in the stack or register state for the entirety of this frame since it's being returned?

Reply:
Yeah, I started to be a bit liberal with `GC_GUARD`...

Reply:
The more I learn about how compilers work the more scared I am that on day some compiler will do something really clever like store a pointer to the middle of a struct in a register, rather than the actual VALUE, and break everything :/

I wonder if anybody has tried compiling Ruby with LTO, maybe aggressive enough cross-TU inlining could make this happen…